snips-nlu-lib/src/intent_classifier/featurizer.rs

```diff
@@ -113,37 +145,45 @@ fn get_word_cluster_features<C: WordClusterer>(
     compute_all_ngrams(&tokens_ref[..], tokens_ref.len())
         .into_iter()
         .filter_map(|ngram| word_clusterer.get_cluster(&ngram.0.to_lowercase()))
-        .collect()
+        .sorted()
 }
 
-fn get_dataset_entities_features<S: Stemmer>(
-    query_tokens: &[String],
-    stemmer: Option<&S>,
+fn get_dataset_entities_features(
+    normalized_stemmed_tokens: &[String],
     entity_utterances_to_feature_names: &HashMap<String, Vec<String>>,
 ) -> Vec<String> {
-    let normalized_tokens: Vec<String> = query_tokens.iter().map(|t| normalize(t)).collect();
-    let normalized_stemmed_tokens = stemmer.map_or(normalized_tokens.clone(), |stemmer| {
-        normalized_tokens
-            .into_iter()
-            .map(|t| stem(&t, stemmer))
-            .collect()
-    });
     let tokens_ref = normalized_stemmed_tokens.iter().map(|t| &**t).collect_vec();
     compute_all_ngrams(&*tokens_ref, tokens_ref.len())
         .into_iter()
         .filter_map(|ngrams| entity_utterances_to_feature_names.get(&ngrams.0))
         .flat_map(|features| features)
         .map(|s| normalize(s))
-        .collect()
+        .sorted()
 }
 
-fn stem<S: Stemmer>(input: &str, stemmer: &S) -> String {
-    stemmer.stem(input)
+fn normalize_stem<S: Stemmer>(tokens: Vec<String>, opt_stemmer: Option<S>) -> Vec<String> {
+    opt_stemmer
+        .map(|stemmer| tokens.iter().map(|t| stemmer.stem(&normalize(t))).collect())
+        .unwrap_or(tokens.iter().map(|t| normalize(t)).collect())
```

Comment:
Use `unwrap_or_else` instead to avoid to evaluate the parameter before we actually need it